src/SignalHandler.cc

```diff
@@ -31,13 +34,52 @@ using namespace gz;
 using namespace common;
 
 // A wrapper for the sigaction sa_handler.
+// TODO(azeey) We should avoid using objects with non-trivial destructors as globals.
 GZ_COMMON_VISIBLE std::map<int, std::function<void(int)>> gOnSignalWrappers;
 std::mutex gWrapperMutex;
 
+namespace
+{
+
+class SelfPipe
+{
+  public:
+  static int pipeFd[2];
+
+  public:
+  static void Initialize();
+
+  public:
+  ~SelfPipe();
+
+  private:
+  SelfPipe();
+
+  private:
+  void CheckPipe();
+
+  private:
+  std::thread checkPipeThread;
+
+  private:
+  std::atomic<bool> running{false};
+};
+
+int SelfPipe::pipeFd[2];
+
+void onSignalTopHalf(int _value)
+{
+  auto valueByte = static_cast<std::uint8_t>(_value);
+  if (write(SelfPipe::pipeFd[1], &valueByte, 1) == -1)
+  {
+    // TODO (azeey) Not clear what to do here.
+  }
+}
+
 /////////////////////////////////////////////////
 /// \brief Callback to execute when a signal is received.
 /// \param[in] _value Signal number.
-void onSignal(int _value)
+void onSignalBottomHalf(int _value)
```

Comment:
Moved the code into the thread since that's the only place it's used. I think it's clearer. https://github.com/gazebosim/gz-common/pull/618/commits/258a565ac1d22810cbcc92def29bcd5cdaceded9